src/dev/gates.rs

```diff
@@ -0,0 +1,227 @@
+
+use ff::PrimeField;
+
+use crate::plonk::{Circuit, ConstraintSystem};
+
+#[derive(Debug)]
+struct Constraint {
+    name: &'static str,
+    expression: String,
+}
+
+#[derive(Debug)]
+struct Gate {
+    name: &'static str,
+    constraints: Vec<Constraint>,
+}
+
+/// A struct for collecting and displaying the gates within a circuit.
+///
+/// # Examples
+///
+/// ```
+/// use ff::Field;
+/// use halo2::{
+///     circuit::{Layouter, SimpleFloorPlanner},
+///     dev::CircuitGates,
+///     plonk::{Circuit, ConstraintSystem, Error},
+///     poly::Rotation,
+/// };
+/// use pasta_curves::pallas;
+///
+/// #[derive(Copy, Clone)]
+/// struct MyConfig {}
+///
+/// #[derive(Clone, Default)]
+/// struct MyCircuit {}
+///
+/// impl<F: Field> Circuit<F> for MyCircuit {
+///     type Config = MyConfig;
+///     type FloorPlanner = SimpleFloorPlanner;
+///
+///     fn without_witnesses(&self) -> Self {
+///         Self::default()
+///     }
+///
+///     fn configure(meta: &mut ConstraintSystem<F>) -> MyConfig {
+///         let a = meta.advice_column();
+///         let b = meta.advice_column();
+///         let c = meta.advice_column();
+///         let s = meta.selector();
+///
+///         meta.create_gate("R1CS constraint", |meta| {
+///             let a = meta.query_advice(a, Rotation::cur());
+///             let b = meta.query_advice(b, Rotation::cur());
+///             let c = meta.query_advice(c, Rotation::cur());
+///             let s = meta.query_selector(s);
+///
+///             Some(("R1CS", s * (a * b - c)))
+///         });
+///
+///         // We aren't using this circuit for anything in this example.
+///         MyConfig {}
+///     }
+///
+///     fn synthesize(&self, _: MyConfig, _: impl Layouter<F>) -> Result<(), Error> {
+///         // Gates are known at configure time; it doesn't matter how we use them.
+///         Ok(())
+///     }
+/// }
+///
+/// let gates = CircuitGates::collect::<pallas::Base, MyCircuit>();
+/// assert_eq!(
+///     format!("{}", gates),
+///     r#####"R1CS constraint:
+/// - R1CS:
+///   S0 * (A0@0 * A1@0 - A2@0)
+/// Total gates: 1
+/// Total custom constraint polynomials: 1
+/// Total negations: 1
+/// Total additions: 1
+/// Total multiplications: 2
+/// "#####,
+/// );
+/// ```
+#[derive(Debug)]
+pub struct CircuitGates {
+    gates: Vec<Gate>,
+    total_negations: usize,
+    total_additions: usize,
+    total_multiplications: usize,
+}
+
+impl CircuitGates {
+    /// Collects the gates from within the circuit.
+    pub fn collect<F: PrimeField, C: Circuit<F>>() -> Self {
+        // Collect the graph details.
+        let mut cs = ConstraintSystem::default();
+        let _ = C::configure(&mut cs);
+
+        let format_scalar = |s: F| {
+            // Format scalar as hex.
+            let s = format!("{:?}", s);
+            // Remove leading zeroes.
+            let s = s.strip_prefix("0x").unwrap();
+            let s = s.trim_start_matches('0');
+            format!("0x{}", s)
```

Comment:
This formats 0 as just "0x".

Reply:
We never reach here with `0`; that constant is handled explicitly below.

Reply:
Hmm, actually we'd reach here if we multiplied an expression by the scalar 0, so we should fix this. But that seems more indicative of a circuit bug.